sourcegraph/sourcegraph
# cody: check whether email is verified when accessing cody

This fixes #51453 by ensuring that on sourcegraph.com users can only access the Cody-relevant endpoints/resolvers if they have >1 verified email.

It does this by 
- adding a helper to the `cody` package
- changing the endpoints/resolvers to check whether verified-email-is-required and user-has-verified-email

This also adds two new GraphQL fields:

```graphql
user {
 # [...]
 hasVerifiedEmail
}
site {
 # [...]
 requiresVerifiedEmailForCody
}
```

Both of them can then be used in clients (VS Code, web) to check whether `site.requiresVerifiedEmailForCody && user.hasVerifiedEmail`.

They should check, otherwise the user will run into an error. Client changes will be made in follow-up PRs.

## Test plan

Existing and new unit tests

#### Prep

1. Create user: `sg db add-user --username test-user-1 --password 123456789101112`
1. Unset site-admin bit: `psql -c "update users set site_admin = false where username = 'test-user-1';"`
1. Unverify email: `psql -c "update user_emails set verified_at = NULL where email = 'test-user-1@sourcegraph.com';"`
1. Login with user and create access token

#### Test

1. Start sourcegraph in dotcom mode: `sg start dotcom`
2. Confirm that all endpoints error:

    Completions stream:
    ```
    curl 'https://sourcegraph.test:3443/.api/completions/stream' \
      -H 'Authorization: token sgp_b3fc2922cb697080210467ac1e727d8d936b0afc' \
      --data-raw $'{"messages":[{"speaker":"human","text":"Testing"}],"temperature":0.2,"maxTokensToSample":1000,"topK":-1,"topP":-1}' \
      --compressed
    ```
    -> prints `cody requires a verified email address`

    Code completions:
    ```
    curl 'https://sourcegraph.test:3443/.api/completions/code' \
      -H 'Authorization: token sgp_b3fc2922cb697080210467ac1e727d8d936b0afc' \
      --data-raw $'{"messages":[{"speaker":"human","text":"Testing"}],"temperature":0.2,"maxTokensToSample":1000,"topK":-1,"topP":-1}' \
      --compressed

    ```
    -> prints `cody requires a verified email address`

    GraphQL completions:
    ```
    curl 'https://sourcegraph.test:3443/.api/graphql' \
      -H 'Authorization: token sgp_b3fc2922cb697080210467ac1e727d8d936b0afc' \
      --data-raw $'{"query":"query($input:CompletionsInput\u0021) {\\n  completions(input: $input)\\n}","variables":{"input":{"messages":[{"speaker":"HUMAN","text":"hello"}],"temperature":0.2,"maxTokensToSample":1000,"topK":-1,"topP":-1}}}' \
      --compressed
    ```
    -> prints `{"errors":[{"message":"cody requires a verified email address","path":["completions"]}],"data":null}`


3. Verify email: `psql -c "update user_emails set verified_at = now() where email = 'test-user-1@sourcegraph.com';"`
4. Retry requests: all work
